### PR TITLE
[feat] Support cache ListFiles result cache in session level

### LINKED ISSUE #7618: [Cache] Support cache ListFiles result cache in session level
### Is your feature request related to a problem or challenge?

In our system, sometimes will call https://github.com/apache/arrow-datafusion/blob/6a775b761a544a8c8079471733290f23a3d62861/datafusion/core/src/datasource/listing/url.rs#L188 to get all files on the remote storage path.  It's unstable between 100ms~2s （list thousands of files）

### Describe the solution you'd like

Support result cache in this functions like spark：

https://github.com/apache/spark/blame/1ccb06ca8cf439e0c13ffbfb50365402e7d1330d/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileStatusCache.scala#L64-L72
```
* A cache of the leaf files of partition directories. We cache these files in order to speed
 * up iterated queries over the same set of partitions. Otherwise, each query would have to
 * hit remote storage in order to gather file statistics for physical planning.
 * 
```

So I want to add ListFilesCache under
https://github.com/apache/arrow-datafusion/blob/83ba66abac5c2bbc4d847a00d3a4ca3c06aadae6/datafusion/execution/src/cache/cache_manager.rs#L39-L42
User can customized their cache policy
make it share in session state and default turn off.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Ted-Jiang**
## Which issue does this PR close?

Closes #7618.

### CODE REVIEW

File: datafusion/core/tests/parquet/file_statistics.rs
Comment (Ted-Jiang) [inline datafusion/core/tests/parquet/file_statistics.rs:157]: Here infer_schema will call list_files as well, so create a new Session state here

File: datafusion/execution/src/cache/cache_unit.rs
Comment (Ted-Jiang) [inline datafusion/execution/src/cache/cache_unit.rs:98]: I think here keep simple, user can use their own notify system to decide call `remove` or `clear`

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1408]: I am worried that providing a Default for SessionState might allow subtle bugs, specifically accidentally running queries or other operations without properly threading through the configuration or runtime. I believe that is why there is no `SessionState::new()` or `Default` 

🤔  this context would be nice to add as a comment to SessonState

File: datafusion/execution/src/cache/cache_unit.rs
Comment (alamb) [inline datafusion/execution/src/cache/cache_unit.rs:101]: the use of `statistics` as a name here seems confusing as this doesn't hold statistics, but rather paths.

Perhaps a name like `paths` would be clearer 🤔

File: datafusion/execution/src/cache/cache_manager.rs
Comment (alamb) [inline datafusion/execution/src/cache/cache_manager.rs:82]:
```suggestion
    /// Enable cache of files to list when listing files.
```

**Review by alamb — COMMENTED** [review]
This looks good to me @Ted-Jiang  -- thank you. The only thing I think needs to be addressed prior to merge is the addition of `SessionState::default`

**Comment by alamb** [comment]
I was thinking a little about the cache APIs here -- at the moment these are all related to `ListingTable` but are attached to the `SessionState`. I wonder if we should make a containing structure like `ListingTableCache` or something to contain them 🤔 .

This is not a suggested change for this PR, just something I was thinking about.

File: datafusion/core/src/execution/context.rs
Comment (Ted-Jiang) [reply datafusion/core/src/execution/context.rs:1408]: Yes, it there are many options to set in `SessionState `,  i just want to write less code when construct SessionState see there are already default in `SessionConfig` and `RuntimeEnv` 😂 ， it make sense  let user point what conf they want.

**Comment by Ted-Jiang** [comment]
> I was thinking a little about the cache APIs here -- at the moment these are all related to `ListingTable` but are attached to the `SessionState`. I wonder if we should make a containing structure like `ListingTableCache` or something to contain them 🤔 .
> 
> This is not a suggested change for this PR, just something I was thinking about.

Thanks for your kindly review! Yes the cache method here are all in `ListingTable`,  I will check refine it  later. 

As this is cache  , it need share the cache result in some place,  i think there is only `SessionState` has the state can share in some level like(session), is there any other way to store the result ?

**Comment by suremarc** [comment]
Are there any advantages to caching file listing in DataFusion instead of inside the `ObjectStore`? It seems to me it should be possible to implement this without any changes to DataFusion itself. There may be tradeoffs there I'm not aware of though

File: datafusion/core/src/execution/context.rs
Comment (alamb) [reply datafusion/core/src/execution/context.rs:1408]: I think we should remove the `Default` impl and instead have a function for tests (`test_session_state()` or similar) to avoid misuse of this API. 

Here is a PR to add some comments about the rationale for not having a `Default` impl:  https://github.com/apache/arrow-datafusion/pull/7654

**Review by alamb — COMMENTED** [review]
Thanks @Ted-Jiang  -- I think we should remove the `Default` impl for `SessionState` from this PR (though if other maintainers have a different opinion I would be happy to defer to the consensus). 

Otherwise I think this PR looks ready to go.

**Comment by alamb** [comment]
>  Are there any advantages to caching file listing in DataFusion instead of inside the ObjectStore? It seems to me it should be possible to implement this without any changes to DataFusion itself. There may be tradeoffs there I'm not aware of though


This is an excellent point @suremarc  -- what do you think @Ted-Jiang  of making a Caching `ObjectStore` implementation? As @suremarc  points out, there is nothing DataFusion specific about this feature, so perhaps it would be better modeled at the `ObjectStore` level 🤔

**Comment by alamb** [comment]
Converting to draft as it is not waiting on a review -- we are now working through comments

File: datafusion/core/src/test/object_store.rs
Comment (Ted-Jiang) [inline datafusion/core/src/test/object_store.rs:48]: Remove the default of SeesionState

**Comment by Ted-Jiang** [comment]
@alamb  @suremarc Sorry for the delay. 

|  making a Caching ObjectStore implementation?

As i have check the some remote store client like `hdfs` there are no cache result in client side,  IMO doing this is `ObjectStore` may be difficult define the work scope (ensure correctness), in datafusion we can support in in session level to not effect others 🤔

**Comment by Ted-Jiang** [comment]
@alamb could you take a look plz

**Comment by suremarc** [comment]
> @alamb @suremarc Sorry for the delay.
> 
> > making a Caching ObjectStore implementation?
> 
> As i have check the some remote store client like `hdfs` there are no cache result in client side, IMO doing this is `ObjectStore` may be difficult define the work scope (ensure correctness), in datafusion we can support in in session level to not effect others 🤔

I agree it would be more difficult, having implemented some caching of my own recently. A full implementation would require dealing with invalidation/consistency, particularly for update operations, as well as supporting nontrivial APIs like `list_with_delimiter`. It would be nice if we could do it there but it's unclear to me if it can be done perfectly.

On that note, I'd like to point out that this cache implementation won't avoid hitting object storage if your table has partition columns, because in that case DataFusion will instead call `list_with_delimiter` recursively, starting from the top-level path. It would still be an improvement, but I am not sure if this will speed up query planning as much as you want. Just wanted to make sure you're aware of this, as I previously had thought that simply caching `list_all_files` would avoid hitting object storage altogether.

File: datafusion/execution/src/cache/cache_manager.rs
Comment (alamb) [inline datafusion/execution/src/cache/cache_manager.rs:84]:
```suggestion
    /// Enable cache of file metadata when listing files.
    /// This setting avoids listing file meta of the same path repeatedly 
    /// in same session, which may be expensive in certain situations (e.g. remote object storage). 
    /// Note that if this option is enabled, DataFusion will not see any updates to the underlying
    /// location.  
    /// Default is disable.
```

**Review by alamb — APPROVED** [review]
Thanks @Ted-Jiang  and @suremarc 

Do you have any performance measurements you can shar @Ted-Jiang  about how much this feature increases performance for your usecase?

**Comment by Ted-Jiang** [comment]
> Thanks @Ted-Jiang and @suremarc
> 
> Do you have any performance measurements you can shar @Ted-Jiang about how much this feature increases performance for your usecase?

We try to fix the situation when calling remote storage list file statistics sometimes are not stable: 
we run one query avg cost 4 seconds, but sometimes it cost double time, we print some log for debug:
```
2023-09-12T06:10:59.098033Z  INFO tokio-runtime-worker ThreadId(16) datafusion::execution::context: Datafusion optimize logical plan cost: 10.852243ms
2023-09-12T06:11:03.309479Z  INFO tokio-runtime-worker ThreadId(16) datafusion::execution::context: Datafusion create physical plan cost: **_4.2222994s_**
2023-09-12T06:11:03.309543Z  INFO tokio-runtime-worker ThreadId(16) ballista_scheduler::planner: planning query stages for job VyQVhat
2023-09-12T06:11:06.439247Z  INFO tokio-runtime-worker ThreadId(20) ballista_scheduler::display: === [VyQVhat/1] Stage finished, physical plan with metrics ===
```
we saw half time cost in create physical plan 🤣  (btw we pass logic plan LIST_TABLE to datafusion in out front end (written in JAVA))

after enable cache we fix this problem the query time keep stable.
btw we have fix the _consistency_ in our java side, because out sys build kind of materialized view, if we we change the source data the path as the cache key also change

**Comment by alamb** [comment]
Thanks again @Ted-Jiang  and @suremarc

**Pull request merged**
